kube/src/client/mod.rs

```diff
@@ -438,7 +471,47 @@ impl TryFrom<Config> for Client {
         let inner = ServiceBuilder::new()
             .layer(common)
             .option_layer(maybe_auth)
-            .layer(tower::layer::layer_fn(LogRequest::new))
+            .layer(
+                // TODO Add OTEL attributes? https://github.com/clux/kube-rs/issues/457
+                // - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md
+                // - https://docs.rs/tracing-opentelemetry/0.13.0/tracing_opentelemetry/#semantic-conventions
+                // - https://docs.rs/tracing-opentelemetry/0.13.0/tracing_opentelemetry/#special-fields
+                TraceLayer::new_for_http()
+                    .make_span_with(|req: &Request<hyper::Body>| {
+                        tracing::debug_span!(
+                            "HTTP",
+                             http.method = %req.method(),
+                             http.uri = %req.uri(),
+                             http.status_code = tracing::field::Empty,
```

Comment:
You already got the required ones there! Although `http.uri` should be `http.url` according to semantic_conventions + otel.

The `otel.kind` + `otel.status_code` should probably be set as well (they are displayed by default in at least some trace viewers, see screenshot linked in other comment).

Reply:
shame there's no way to get the `HTTP` name to be more dynamic. If we got the name of the request call (like `patch_status`) to be the span name, then we could possibly drop the manual `#[instrument]`ations.

Reply:
Yeah, it's possible to derive a name from `Request` and use that with `otel.name` to override the span name:

```rust
TraceLayer::new_for_http()
    .make_span_with(|req: &Request<hyper::Body>| {
        tracing::debug_span!(
            "HTTP",
            http.method = %req.method(),
            http.url = %req.uri(),
            http.status_code = tracing::field::Empty,
            // Override the span name sent to OpenTelemetry exporters.
            otel.name = otel_name(req),
            otel.kind = "client",
            otel.status_code = tracing::field::Empty,
        )
    })
```

But not sure what's good.

We can do `HTTP {METHOD_NAME}` like suggested in the semantic convention.
```rust
#[inline]
fn otel_name(req: &Request<hyper::Body>) -> &'static str {
    match *req.method() {
        http::Method::POST => "HTTP POST",
        http::Method::GET => "HTTP GET",
        http::Method::PUT => "HTTP PUT",
        http::Method::PATCH => "HTTP PATCH",
        http::Method::DELETE => "HTTP DELETE",
        _ => "HTTP",
    }
}
```

> If we got the name of the request call (like `patch_status`)

Maybe insert the name in [extensions](https://docs.rs/http/0.2.4/http/struct.Extensions.html):
```rust
// in Api
req.extensions_mut().insert("patch_status");
```
and
```rust
otel.name = req.extensions().get::<&'static str>().unwrap_or(&"HTTP"),
```
?

Reply:
Oh nice! Extensions is a good idea! Interesting solution they have chosen with the TypeId map rather than explicit keys, but seems to be a memory optimisation for small amount of entries. It's probably okay to use it, it generally stays internal anyway (unless people pick apart it with our `Request` type, but that's also fine).

Reply:
New version looks great. I just tried running through Tempo via controller-rs.

![Screenshot from 2021-06-04 22-21-41](https://user-images.githubusercontent.com/134092/120864345-ffec5c80-c583-11eb-909a-d09765f7c073.png)

Semantic convention looks accurate, extension passed span name is picked up correctly.

Interestingly, if you are in tracing land, then the `debug!("requesting")` is pointless and only inflates the trace, but if you are in stdout land, then you need to have it - as `tracing_subscriber::fmt::init()` only prints events and not spans. Don't think there's anything that can be done about that at the moment though.

Reply:
> as `tracing_subscriber::fmt::init()` only prints events and not spans. 

It's possible to make `tracing_subscriber::fmt` log events on span lifecycle.

```rust
use tracing_subscriber::fmt::format::FmtSpan;

tracing_subscriber::fmt()
    .with_span_events(FmtSpan::CLOSE)
    .init();
```
will log when the span closes.